theano/tensor/basic.py

```diff
@@ -5269,6 +5266,129 @@ def dot(a, b):
 # Linalg : TensorDot
 #########################
 
+def _tensordot_as_dot(a, b, axes, dot, batched):
+    """
+    Reduces a tensor dot product to a matrix or vector dot product. Based
+    on code from Tijmen Tieleman's gnumpy
+    (http://www.cs.toronto.edu/~tijmen/gnumpy.html).
+
+    Please see the documentation of tensordot for the meaning of the a, b
+    and axes arguments.
+
+    :param dot: a function that accepts two symbolic variables and computes
+                the appropriate dot product (e.g. dot, batched_dot)
+    :type dot: function
+
+    :param batched: whether to treat the first axis of a and b as a batch
+                    axis.  If so, this axis will be preserved in the output,
+                    allowing this function to be used also for batched
+                    tensor dot products.
+    :type batched: boolean
+
+    :returns: a tensor with shape equal to the concatenation of a's shape
+              (less any dimensions that were summed over) and b's shape
+              (less the first dimension and any dimensions that were summed
+              over).
+    :rtype: symbolic tensor
+    """
+    a, b = as_tensor_variable(a), as_tensor_variable(b)
+
+    if not numpy.isscalar(axes) and len(axes) != 2:
+        raise ValueError('Axes should be an integer or a '
+                         'list/tuple of len 2 (%s was provided)'
+                         % repr(axes))
+
+    # if 'axes' is a number of axes to multiply and sum over (trailing axes
+    # of a, leading axes of b), we can just reshape and use dot.
+    elif numpy.isscalar(axes):
+        axes = int(axes)
+
+        for operand_name, operand in (("a", a), ("b", b)):
+            if axes > operand.ndim:
+                raise ValueError(
+                    'axes can not be larger than the dimension of %s '
+                    '(%s.ndim=%i, axes=%i)'
+                    % (operand_name, operand_name, operand.ndim, axes))
```

Comment:
You can use another python string fonctionality to make this more readable:

"a_string %(data)s %(data)s %(out)d" %{data="test", out=2}

This give name to the string substitution